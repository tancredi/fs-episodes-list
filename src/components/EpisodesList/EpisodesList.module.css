.episodes,
.characters {
  list-style: none;
  padding: 0;
}

.episode,
.character {
  margin: 0;
  display: block;
}

.episodes {
  perspective: 200px;
}

.episode {
  display: flex;
  align-items: flex-start;
  padding: 0;
  margin-bottom: 20px;
  border-radius: 5px;
  background-color: rgba(0, 0, 0, 0.3);
  backdrop-filter: blur(20px);
  overflow: hidden;
  transition: transform 0.2s;
  perspective-origin: center center;
}

@media (hover: hover) {
  .episode:hover {
    transform: rotate3d(1, 0, 0, 1deg) scale(1.05);
  }
}

.info {
  flex: 1;
  padding: 30px;
  min-width: 180px;
  padding-right: 0;
}

.info h3 {
  margin: 0;
  margin-bottom: 5px;
}

.info h4 {
  margin: 0;
  opacity: 0.7;
  font-weight: 200;
}

.image {
  object-fit: cover;
}

.episode-image {
  width: 200px;
  height: 190px;
}

.characters {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-end;
  padding: 30px;
  padding-left: 0;
  margin-bottom: -10px;
}

.character {
  width: 60px;
  height: 60px;
  margin-left: 10px;
  margin-bottom: 10px;
}

.character-image {
  width: 100%;
  height: 100%;
  border-radius: 100%;
}

@media (max-width: 800px) {
  .episode,
  .info {
    padding: 0;
  }

  .episode {
    flex-direction: column;
  }

  .episode-image {
    width: 100%;
  }

  .info {
    text-align: center;
    width: 100%;
    margin: 20px 0;
  }

  .characters {
    padding: 30px;
    padding-top: 0;
    justify-content: center;
    width: 100%;
    box-sizing: border-box;
  }
}
