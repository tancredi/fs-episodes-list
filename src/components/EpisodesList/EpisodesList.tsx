import type { FunctionComponent } from "react";
import classNames from "classnames";
import type { Episode } from "src/types/episodes";
import styles from "./EpisodesList.module.css";

export const EpisodesList: FunctionComponent<{ episodes: Episode[] }> = ({
  episodes,
}) => (
  <ul className={styles.episodes}>
    {episodes.map((episode) => (
      <EpisodeListItem key={episode.id} episode={episode} />
    ))}
  </ul>
);

export const EpisodeListItem: FunctionComponent<{ episode: Episode }> = ({
  episode,
}) => (
  <li className={styles.episode}>
    <img
      src={episode.img_url}
      className={classNames(styles.image, styles["episode-image"])}
    />

    <div className={styles.info}>
      {episode.name}
      {episode.air_date}
    </div>

    <ul className={styles.characters}>
      {episode.characters.map((character) => (
        <CharacterThumbnail character={character} key={character.id} />
      ))}
    </ul>
  </li>
);

export const CharacterThumbnail: FunctionComponent<{
  character: Episode["characters"][number];
}> = ({ character }) => (
  <li className={styles.character}>
    <img
      className={classNames(styles.image, styles["character-image"])}
      src={character.img_url}
    />
  </li>
);
