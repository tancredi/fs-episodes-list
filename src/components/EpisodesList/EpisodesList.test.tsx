import React from "react";
import { render } from "@testing-library/react";
import { EpisodesList } from "./EpisodesList";
import type { Episode } from "../../types/episodes";

const MOCK_CHARACTER: Episode["characters"][number] = {
  id: 1,
  name: "Jane Foo",
  status: "Alive",
  species: "Human",
  gender: "Male",
  hair: "Blonde",
  alias: ["Janey", "Jaynei"],
  origin: "Earth",
  abilities: ["Foo skill", "Bar skill"],
  img_url: "http://foo.bar/character-image.png",
};

const MOCK_EPISODE: Episode = {
  id: 1,
  name: "Episode foobar",
  air_date: "10/10/2010",
  director: "John Doe",
  writer: "Dan Jane",
  characters: [MOCK_CHARACTER],
  img_url: "http://foo.bar/episode-image.png",
};

describe("Episodes list components", () => {
  it("renders UI correctly", () => {
    expect(
      render(<EpisodesList episodes={[MOCK_EPISODE]} />).asFragment()
    ).toMatchSnapshot();
  });

  it("renders each episode provided", () => {
    const episodes = [
      { ...MOCK_EPISODE, id: 0, name: "Episode foo" },
      { ...MOCK_EPISODE, id: 1, name: "Episode bar" },
    ];
    const { getByText } = render(<EpisodesList episodes={episodes} />);

    getByText("Episode foo");
    getByText("Episode bar");
  });

  it("renders expected episode info", () => {
    const { getByText } = render(<EpisodesList episodes={[MOCK_EPISODE]} />);

    getByText(MOCK_EPISODE.name);
    getByText(MOCK_EPISODE.air_date);
  });

  it("renders the main episodes’ image", () => {
    const image = render(
      <EpisodesList episodes={[MOCK_EPISODE]} />
    ).getByAltText('Image from episode "Episode foobar"');

    expect(image).toHaveAttribute("src", MOCK_EPISODE.img_url);
  });

  it("renders all characters", () => {
    const { getByTitle } = render(
      <EpisodesList
        episodes={[
          {
            ...MOCK_EPISODE,
            characters: [
              { ...MOCK_CHARACTER, id: 0, name: "Jane" },
              { ...MOCK_CHARACTER, id: 1, name: "John" },
            ],
          },
        ]}
      />
    );

    getByTitle("Jane");
    getByTitle("John");
  });

  it("renders each characters’ image", () => {
    const { getByAltText } = render(<EpisodesList episodes={[MOCK_EPISODE]} />);

    expect(getByAltText(`Picture of ${MOCK_CHARACTER.name}`)).toHaveAttribute(
      "src",
      MOCK_CHARACTER.img_url
    );
  });
});
