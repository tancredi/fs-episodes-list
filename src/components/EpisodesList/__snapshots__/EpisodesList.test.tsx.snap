// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episodes list component renders UI correctly 1`] = `
<DocumentFragment>
  <ul
    class="episodes"
  >
    <li
      class="episode"
    >
      <img
        alt="Image from episode "Episode foobar""
        class="image episode-image"
        src="http://foo.bar/episode-image.png"
      />
      <div
        class="info"
      >
        <h3>
          Episode foobar
        </h3>
        <h4>
          10/10/2010
        </h4>
      </div>
      <ul
        class="characters"
      >
        <li
          class="character"
          title="Jane Foo"
        >
          <img
            alt="Picture of Jane Foo"
            class="image character-image"
            src="http://foo.bar/character-image.png"
          />
        </li>
      </ul>
    </li>
  </ul>
</DocumentFragment>
`;
